src/jsifier.js

```diff
@@ -130,16 +130,15 @@ function JSify(data, functionsOnly) {
   }
 
   // Generates a function that invokes a proxied function call from the calling thread to the main browser thread.
-  function generateProxiedCallInvoker(sig, sync /*async if false*/) {
+  function generateProxiedCallInvoker(sig) {
     if (sig.length == 0) throw 'Function signature cannot be empty!';
     function argsList(num) { // ", p0, p1, p2, p3, p4"
       var s = '';
       for(var i = 0; i < num; ++i) s += ', p' + i;
       return s;
     }
 
-    var func = "function _emscripten_" + (sync ? '' : 'a') + 'sync_run_in_browser_thread_' + sig + '(func' + argsList(sig.length-1) + ') {\n';
-    if (sync) func += '  var waitAddress = stackSave();\n';
+    var func = 'function _emscripten_async_run_in_browser_thread_' + sig + '(func' + argsList(sig.length-1) + ') {\n';
```

Comment:
This looks like it changes only the sync proxied functions to go through C/C++ heap, but not the async proxied ones, which still would use postMessage. That unfortunately will not work, because we need to guarantee sequential consistency between all calls. If sync functions go through different pipe than async functions, then it is possible that the order in which they are executed can change, which can be catastrophical. (e.g. some WebGL functions are sync, others async, but they must all occur in the same sequential order that they are dispatched).

I think to resolve this, also the async functions can be switched over to go through the C/C++ side proxying pipe.

Reply:
Thanks! I didn't realize that, and thought it was separable. I'll move them all to postMessage in this single PR, then.